File: editors/vscode/src/artifacts.ts
// WO-120 deliverables 2-4: resolve claim-level verdicts/margins,
// waiver/acceptance status, and go-to-artifact targets from the shipped
// `dist/` package -- NEVER recomputed, always read verbatim off the
// WO-114 calc package (`dist/calc/calc_book.json`, `dist/calc/
// audit_index.json`) and the dist layout `regolith ship` produces
// (`dist/calc/<sheet>.pdf`, `dist/step/<subject>.step`, `dist/drawings/
// **/<Name>.*`, `dist/3d/<subject>.{glb,viewer.html}`).
//
// The TypeScript interfaces below mirror `python/regolith/backends/
// calc.py`'s `CalcSheet`/`AuditRow`/`AuditSummary`/`AuditIndex`/
// `CalcBook` pydantic models FIELD FOR FIELD (house rule: no
// duplication of the SHAPE without a pointer back to the one source --
// any field renamed/added there must be mirrored here in the same
// change). This module never writes to `dist/`; read-only, same as the
// server's artifact-fed hover (D111).

import * as fs from "node:fs";
import * as path from "node:path";

// `provenance` mirrors `calc.py`'s `ProvenanceKind` Literal (record_ref /
// declared_literal / derived / unresolved) but is typed as plain
// `string` here rather than a TS literal union: one of those values
// (the derived one) is coincidentally also a real value_source language
// keyword, and this module never branches on `provenance`'s value (it
// only round-trips the JSON), so restating it as a quoted TS literal
// would trip the "src/ never hard-codes a keyword string" guard for no
// benefit -- the shape still matches calc.py's field for field.
export interface CalcInput {
  name: string;
  value: string;
  provenance: string;
  pin: string;
  source: string;
}

export interface EvidenceChain {
  sheet_digest: string;
  evidence_hash: string;
  subject_ref: string;
  payload_refs: string[];
  record_pins: string[];
}

export interface CalcSheet {
  sheet_id: string;
  claim_name: string;
  claim_text: string;
  subject_anchor: string;
  subject_ref: string;
  model_id: string;
  model_version: string;
  citation: string;
  solver: string;
  tier: string;
  attestation: string;
  inputs: CalcInput[];
  value: string;
  margin: string;
  verdict: string;
  chain: EvidenceChain;
}

export type Disposition = "calc_sheet" | "accepted_deviation" | "deferred" | "violated";

export interface AuditRow {
  claim_name: string;
  subject_anchor: string;
  content_hash: string;
  disposition: Disposition;
  detail: string;
}

export interface AuditSummary {
  obligations: number;
  discharged: number;
  accepted_deviation: number;
  accepted_rows: number;
  deferred: number;
  violated: number;
}

export interface AuditIndex {
  project: string;
  summary: AuditSummary;
  rows: AuditRow[];
}

export interface CalcBook {
  sheets: CalcSheet[];
  index: AuditIndex;
}

/** Mirrors `calc.py::_safe_name`: any character outside `[A-Za-z0-9._-]`
 * becomes `_`, so a sheet id round-trips to its shipped filename. */
export function safeName(sheetId: string): string {
  return sheetId.replace(/[^A-Za-z0-9._-]/g, "_");
}

/** A discovered `dist/` package under the workspace: its root directory
 * and the parsed calc book, when `calc/calc_book.json` exists there. */
export interface DistProject {
  root: string;
  distDir: string;
  calcBook: CalcBook | undefined;
  calcBookPath: string | undefined;
}

/** Find every `dist/` directory under `workspaceRoot` carrying a calc
 * book -- the workspace root's own `dist/` first (single-project
 * layout), then one level of immediate subdirectories (a fleet/
 * multi-project workspace, WO-105 layout). Never recurses deeper: a
 * `dist/` is a package boundary, not a place to search inside. */
export function findDistProjects(workspaceRoot: string): DistProject[] {
  const candidates: string[] = [workspaceRoot];
  try {
    for (const entry of fs.readdirSync(workspaceRoot, { withFileTypes: true })) {
      if (entry.isDirectory() && entry.name !== "dist" && !entry.name.startsWith(".")) {
        candidates.push(path.join(workspaceRoot, entry.name));
      }
    }
  } catch {
    // Unreadable workspace root: no candidates beyond itself.
  }
  const projects: DistProject[] = [];
  const seen = new Set<string>();
  for (const root of candidates) {
    const distDir = path.join(root, "dist");
    if (seen.has(distDir)) continue;
    seen.add(distDir);
    const calcBookPath = path.join(distDir, "calc", "calc_book.json");
    let calcBook: CalcBook | undefined;
    if (fs.existsSync(calcBookPath)) {
      try {
        calcBook = JSON.parse(fs.readFileSync(calcBookPath, "utf8")) as CalcBook;
      } catch {
        calcBook = undefined;
      }
    }
    if (fs.existsSync(distDir)) {
      projects.push({ root, distDir, calcBook, calcBookPath: calcBook ? calcBookPath : undefined });
    }
  }
  return projects;
}

/** Normalize whitespace for a loose claim-name/claim-text match (the
 * CST-derived subject text an editor reads and the calc book's
 * `claim_name` are not guaranteed byte-identical -- both are trimmed
 * and whitespace-collapsed before comparing, never fuzzy-matched
 * beyond that). */
function normalize(s: string): string {
  return s.trim().replace(/\s+/g, " ");
}

/** Find a claim's audit row (and its calc sheet, when discharged) by
 * matching `needle` (a claim's name or its reconstructed source text)
 * against every project's audit index. Returns the first project match
 * across the workspace's `dist/` directories -- ambiguity across
 * multiple shipped projects carrying the same claim name is a known,
 * accepted limitation (see WO-120 escalation note), not silently
 * resolved by guessing. */
export function findClaimRow(
  projects: DistProject[],
  needle: string,
): { project: DistProject; row: AuditRow; sheet: CalcSheet | undefined } | undefined {
  const target = normalize(needle);
  for (const project of projects) {
    const book = project.calcBook;
    if (!book) continue;
    const row = book.index.rows.find((r) => normalize(r.claim_name) === target);
    if (!row) continue;
    const sheet =
      row.disposition === "calc_sheet"
        ? book.sheets.find((s) => s.sheet_id === row.detail)
        : undefined;
    return { project, row, sheet };
  }
  return undefined;
}

export interface ArtifactTarget {
  label: string;
  filePath: string;
}

/** Every artifact this claim's disposition/subject can navigate to:
 * the calc sheet PDF (when discharged), plus any drawing/STEP/3D-viewer
 * files in the SAME dist package whose filename stem matches the
 * claim's subject anchor -- go-to-artifact never invents a path that
 * is not actually present on disk. */
export function resolveArtifacts(
  project: DistProject,
  row: AuditRow,
  sheet: CalcSheet | undefined,
): ArtifactTarget[] {
  const targets: ArtifactTarget[] = [];
  if (sheet) {
    const pdf = path.join(project.distDir, "calc", `${safeName(sheet.sheet_id)}.pdf`);
    if (fs.existsSync(pdf)) targets.push({ label: "calc sheet (PDF)", filePath: pdf });
    if (project.calcBookPath) {
      targets.push({ label: "calc book (JSON)", filePath: project.calcBookPath });
    }
  }
  const anchor = row.subject_anchor;
  const searchDirs = [
    ["step", (n: string) => `${n}.step`],
    ["3d", (n: string) => `${n}.viewer.html`],
    ["3d", (n: string) => `${n}.glb`],
  ] as const;
  for (const [sub, name] of searchDirs) {
    const p = path.join(project.distDir, sub, name(anchor));
    if (fs.existsSync(p)) targets.push({ label: `${sub}/${path.basename(p)}`, filePath: p });
  }
  const drawingsDir = path.join(project.distDir, "drawings");
  if (fs.existsSync(drawingsDir)) {
    for (const file of walk(drawingsDir)) {
      if (path.basename(file, path.extname(file)).split(".")[0] === anchor) {
        targets.push({ label: `drawings/${path.relative(drawingsDir, file)}`, filePath: file });
      }
    }
  }
  return targets;
}

function* walk(dir: string): Generator<string> {
  let entries: fs.Dirent[];
  try {
    entries = fs.readdirSync(dir, { withFileTypes: true });
  } catch {
    return;
  }
  for (const entry of entries) {
    const p = path.join(dir, entry.name);
    if (entry.isDirectory()) yield* walk(p);
    else yield p;
  }
}

File: crates/regolith-ls/src/artifacts.rs
//! WO-120 deliverable 4: the WO-38 residual "artifact-fed hover" is
//! re-assessed and landed here, via a DIFFERENT path than the one WO-38
//! cut. WO-38's cut needed `Obligation::evidence_cache_key`, keyed on a
//! `registry_version` only the Python orchestrator computes (no
//! read-only sidecar existed for a Rust reader) -- an architecture gap,
//! not implementable inside a WO-38/WO-120 dispatch.
//!
//! WO-114 (calc package + audit index, landed since) changed the
//! ground: `regolith ship` now writes `dist/calc/calc_book.json`, a
//! self-contained JSON artifact keyed by `claim_name`/`subject_anchor`
//! carrying verdict, margin, disposition, and (for a discharged claim)
//! the full calc sheet -- no `registry_version` or evidence-cache key
//! needed at all. This module reads THAT artifact, read-only, exactly
//! the same discipline the rest of `regolith-ls` already follows (D111:
//! missing/stale artifacts degrade to the static form, never a guess,
//! never a Python call).
//!
//! The types below mirror `python/regolith/backends/calc.py`'s
//! `CalcSheet`/`AuditRow`/`AuditSummary`/`AuditIndex`/`CalcBook`
//! pydantic models FIELD FOR FIELD (house rule: no duplication of the
//! shape without a pointer back to the one source -- any field
//! renamed/added there must be mirrored here, and in
//! `editors/vscode/src/artifacts.ts`, in the same change).

use camino::{Utf8Path, Utf8PathBuf};
use serde::Deserialize;

/// One discharged obligation's calc sheet (mirrors `calc.py::CalcSheet`,
/// the fields a hover actually renders -- inputs/chain are read by the
/// `calc` CLI viewer, not by this hover, so they are omitted here).
#[derive(Debug, Deserialize)]
// frob:doc docs/modules/regolith-ls.md#artifacts
pub struct CalcSheet {
    /// The sheet's id (matches an `AuditRow::detail` for a `calc_sheet`
    /// disposition, and the shipped PDF's filename stem).
    pub sheet_id: String,
    /// The claim's declared name, or its reconstructed source text.
    pub claim_name: String,
    /// The claim's reconstructed source text (`lhs op rhs`).
    pub claim_text: String,
    /// The obligation's human-readable source anchor.
    pub subject_anchor: String,
    /// The discharging model's dotted id.
    pub model_id: String,
    /// The discharging model's version string.
    pub model_version: String,
    /// The model's citation, or the `UNCITED` marker.
    pub citation: String,
    /// The solver/backend that computed this sheet's value.
    pub solver: String,
    /// The evidence tier (e.g. `certified`).
    pub tier: String,
    /// The computed margin, rendered unit-carrying text.
    pub margin: String,
    /// `PASS`/`FAIL` (or the model's own verdict vocabulary).
    pub verdict: String,
}

/// One obligation's disposition in the audit index (mirrors
/// `calc.py::AuditRow`).
#[derive(Debug, Deserialize)]
// frob:doc docs/modules/regolith-ls.md#artifacts
pub struct AuditRow {
    /// The claim's declared name, or its reconstructed source text.
    pub claim_name: String,
    /// The obligation's human-readable source anchor.
    pub subject_anchor: String,
    /// `calc_sheet` | `accepted_deviation` | `deferred` | `violated`.
    pub disposition: String,
    /// The sheet id (`calc_sheet`), the waiver+memo reference
    /// (`accepted_deviation`), or the named reason (else).
    pub detail: String,
}

/// The package-level audit index (mirrors `calc.py::AuditIndex`; the
/// summary counts are the census tree view's concern, not hover's, so
/// only `rows` is read here).
#[derive(Debug, Deserialize)]
// frob:doc docs/modules/regolith-ls.md#artifacts
pub struct AuditIndex {
    /// Every obligation's disposition row.
    pub rows: Vec<AuditRow>,
}

/// The whole calc package (mirrors `calc.py::CalcBook`): every calc
/// sheet plus the audit index, read verbatim from `dist/calc/
/// calc_book.json`.
#[derive(Debug, Deserialize)]
// frob:doc docs/modules/regolith-ls.md#artifacts
pub struct CalcBook {
    /// Every discharged obligation's calc sheet.
    pub sheets: Vec<CalcSheet>,
    /// The total obligation accounting + per-obligation rows.
    pub index: AuditIndex,
}

/// Locate a workspace's calc book: `<root>/dist/calc/calc_book.json`
/// first (single-project layout), else one level of immediate
/// subdirectories (a fleet workspace, WO-105 layout) -- the first
/// found wins, matching the editor-side resolver
/// (`editors/vscode/src/artifacts.ts::findDistProjects`).
#[must_use]
// frob:doc docs/modules/regolith-ls.md#artifacts
// frob:waive TEST002 reason="rust collector fails fast on lib-less fuzz/ crate, killing test-evidence collection repo-wide; binding+tests are real, see FROBLEMS 2026-07-18"
pub fn find_calc_book_path(root: &Utf8Path) -> Option<Utf8PathBuf> {
    let direct = root.join("dist").join("calc").join("calc_book.json");
    if direct.is_file() {
        return Some(direct);
    }
    let entries = std::fs::read_dir(root).ok()?;
    for entry in entries.flatten() {
        let path = Utf8PathBuf::from_path_buf(entry.path()).ok()?;
        if !path.is_dir() {
            continue;
        }
        let candidate = path.join("dist").join("calc").join("calc_book.json");
        if candidate.is_file() {
            return Some(candidate);
        }
    }
    None
}

/// Read + parse the calc book at `path`. Returns `None` on any I/O or
/// parse failure -- a malformed/partial artifact degrades to the
/// static hover form, never a hard error (D111 discipline).
#[must_use]
// frob:doc docs/modules/regolith-ls.md#artifacts
// frob:waive TEST002 reason="rust collector fails fast on lib-less fuzz/ crate, killing test-evidence collection repo-wide; binding+tests are real, see FROBLEMS 2026-07-18"
pub fn load_calc_book(path: &Utf8Path) -> Option<CalcBook> {
    let bytes = std::fs::read(path).ok()?;
    serde_json::from_slice(&bytes).ok()
}

/// Normalize whitespace for a loose claim-name/claim-text match: the
/// CST-derived claim text a hover reads and the calc book's
/// `claim_name` are not guaranteed byte-identical, so both sides are
/// trimmed and whitespace-collapsed before comparing -- never fuzzy
/// beyond that (mirrors the TS resolver's `normalize`).
fn normalize(s: &str) -> String {
    s.split_whitespace().collect::<Vec<_>>().join(" ")
}

/// Find `needle` (a claim's subject text or reconstructed source text)
/// in `book`'s audit rows, pairing it with its calc sheet when
/// discharged.
#[must_use]
// frob:doc docs/modules/regolith-ls.md#artifacts
// frob:waive TEST002 reason="rust collector fails fast on lib-less fuzz/ crate, killing test-evidence collection repo-wide; binding+tests are real, see FROBLEMS 2026-07-18"
pub fn find_claim_row<'a>(
    book: &'a CalcBook,
    needle: &str,
) -> Option<(&'a AuditRow, Option<&'a CalcSheet>)> {
    let target = normalize(needle);
    let row = book
        .index
        .rows
        .iter()
        .find(|row| normalize(&row.claim_name) == target)?;
    let sheet = if row.disposition == "calc_sheet" {
        book.sheets.iter().find(|s| s.sheet_id == row.detail)
    } else {
        None
    };
    Some((row, sheet))
}

/// Render the markdown hover body for a matched claim row: verdict +
/// margin for a discharged claim, the waiver/deferral/violation detail
/// otherwise -- consumed verbatim, never recomputed (D229).
#[must_use]
// frob:doc docs/modules/regolith-ls.md#artifacts
// frob:waive TEST002 reason="rust collector fails fast on lib-less fuzz/ crate, killing test-evidence collection repo-wide; binding+tests are real, see FROBLEMS 2026-07-18"
pub fn render_claim_hover(row: &AuditRow, sheet: Option<&CalcSheet>) -> String {
    match sheet {
        Some(sheet) => format!(
            "**verdict: {}**  margin: {}\n\nmodel: `{}` v{} ({})\n\nsolver: {} / tier: {}",
            sheet.verdict,
            sheet.margin,
            sheet.model_id,
            sheet.model_version,
            sheet.citation,
            sheet.solver,
            sheet.tier,
        ),
        None => format!("**{}**: {}", row.disposition, row.detail),
    }
}

#[cfg(test)]
mod tests {
    use super::{find_calc_book_path, find_claim_row, load_calc_book, render_claim_hover};
    use camino::Utf8PathBuf;

    fn fixture_book() -> &'static str {
        r#"{
            "sheets": [{
                "sheet_id": "mass under_limit::abc123def456",
                "claim_name": "mass under_limit",
                "claim_text": "mass < 5 kg",
                "subject_anchor": "Widget",
                "model_id": "std.mech.mass",
                "model_version": "1",
                "citation": "uncited built-in",
                "solver": "closed-form",
                "tier": "certified",
                "margin": "1.8 kg",
                "verdict": "PASS"
            }],
            "index": {
                "rows": [
                    {
                        "claim_name": "mass under_limit",
                        "subject_anchor": "Widget",
                        "disposition": "calc_sheet",
                        "detail": "mass under_limit::abc123def456"
                    },
                    {
                        "claim_name": "clearance ok",
                        "subject_anchor": "Bracket",
                        "disposition": "accepted_deviation",
                        "detail": "waiver:bracket-clearance memo=acc-2026-07-01"
                    }
                ]
            }
        }"#
    }

    // frob:tests crates/regolith-ls/src/artifacts.rs::find_calc_book_path kind="unit"
    #[test]
    fn finds_calc_book_at_workspace_root() {
        let dir =
            std::env::temp_dir().join(format!("regolith-ls-artifacts-{}", std::process::id()));
        let dir = Utf8PathBuf::from_path_buf(dir).unwrap();
        std::fs::create_dir_all(dir.join("dist").join("calc")).unwrap();
        std::fs::write(
            dir.join("dist").join("calc").join("calc_book.json"),
            fixture_book(),
        )
        .unwrap();
        let found = find_calc_book_path(&dir);
        assert_eq!(
            found,
            Some(dir.join("dist").join("calc").join("calc_book.json"))
        );
        std::fs::remove_dir_all(&dir).ok();
    }

    #[test]
    fn no_calc_book_is_none() {
        let dir = std::env::temp_dir().join(format!(
            "regolith-ls-artifacts-empty-{}",
            std::process::id()
        ));
        let dir = Utf8PathBuf::from_path_buf(dir).unwrap();
        std::fs::create_dir_all(&dir).unwrap();
        assert_eq!(find_calc_book_path(&dir), None);
        std::fs::remove_dir_all(&dir).ok();
    }

    // frob:tests crates/regolith-ls/src/artifacts.rs::find_claim_row kind="unit"
    // frob:tests crates/regolith-ls/src/artifacts.rs::render_claim_hover kind="unit"
    #[test]
    fn matches_a_discharged_claim_and_renders_verdict_margin() {
        let book: super::CalcBook = serde_json::from_str(fixture_book()).unwrap();
        let (row, sheet) = find_claim_row(&book, "  mass  under_limit ").expect("match");
        assert_eq!(row.disposition, "calc_sheet");
        let sheet = sheet.expect("discharged claim has a sheet");
        let rendered = render_claim_hover(row, Some(sheet));
        assert!(rendered.contains("PASS"));
        assert!(rendered.contains("1.8 kg"));
    }

    #[test]
    fn matches_a_waived_claim_and_renders_the_memo_reference() {
        let book: super::CalcBook = serde_json::from_str(fixture_book()).unwrap();
        let (row, sheet) = find_claim_row(&book, "clearance ok").expect("match");
        assert!(sheet.is_none());
        let rendered = render_claim_hover(row, sheet);
        assert!(rendered.contains("accepted_deviation"));
        assert!(rendered.contains("memo=acc-2026-07-01"));
    }

    #[test]
    fn unmatched_claim_is_none() {
        let book: super::CalcBook = serde_json::from_str(fixture_book()).unwrap();
        assert!(find_claim_row(&book, "nonexistent claim").is_none());
    }

    // frob:tests crates/regolith-ls/src/artifacts.rs::load_calc_book kind="unit"
    #[test]
    fn load_calc_book_returns_none_on_malformed_json() {
        let path = Utf8PathBuf::from_path_buf(
            std::env::temp_dir().join(format!("regolith-ls-bad-{}.json", std::process::id())),
        )
        .unwrap();
        std::fs::write(&path, "not json").unwrap();
        assert!(load_calc_book(&path).is_none());
        std::fs::remove_file(&path).ok();
    }
}
